### DISCUSSION TITLE
Segmentation fault in docker container

### DISCUSSION THREAD

**Discussion by uwehuberKEB** [aprimadi/influxdb2#55] [Q&A] [answered]
Hi,
If I use the client in a native linux environment I can access the influx database in a docker container. If I try to start the client in a separate docker container I get a segmentation fault in this call
let client = Client::new("http://myipaddress:8086", "organization", "my_token");

from the docker container I can curl the database. Does anybody has a similar problem or an idea. It would be very helpful.

**Comment by Boudewijn26** [comment]
Hi @uwehuberKEB,

Would it be possible for you to include a debug build in your docker container? That way the segfault should give you a stack trace which would greatly aid troubleshooting

**Comment by Boudewijn26** [comment]
Alternatively you can also use the `ClientBuilder` to instantiate the `Client` instead of `Client::new`. The `ClientBuilder` returns a `Result<Client, BuildError>`, which would also aid troubleshooting

**Reply by uwehuberKEB** [reply]
I used the clientbuilder
![image](https://github.com/aprimadi/influxdb2/assets/76032392/60efbfb2-c575-442e-9816-109f2fe1f7b0)
and put the debug binary into the docker image. Unfortunately I don't get a stack trace
![image](https://github.com/aprimadi/influxdb2/assets/76032392/bdda8688-0b33-4c74-9d6f-78fbbd2aba3b)

The docker image is build with the following file

# Base image for building the application
FROM rust:alpine as builder

# The latest alpine images don't have some tools like (`git` and `bash`).
RUN apk update
RUN apk upgrade
RUN apk add build-base
RUN apk add openssl-dev
RUN apk add musl-dev
RUN apk add pkgconfig
RUN apk add curl
ENV OPENSSL_LIB_DIR=/usr/lib
ENV OPENSSL_INCLUDE_DIR=/usr/include

# Set the Current Working Directory inside the container
WORKDIR /usr/src/app

# copy the complete source for building into the image
COPY . .

# Build the application
RUN cargo build

######## Start a new stage from scratch #######
FROM alpine:latest

ENV RUST_BACKTRACE=1

WORKDIR /root/

# Copy the Pre-built binary file from the previous stage
COPY --from=builder /usr/src/app/target/debug/influxwriter .

# Run the executable
CMD ["./influxwriter"]
 
any other idea?

**Reply by Boudewijn26** [reply]
I don't see the "Client builder created" line in the output. I'm not sure exactly where you are logging "Try to establish influx connection", but it could be breaking before you even get to the `ClientBuilder::with_builder`

I suspect the final docker image might be missing some libraries which are expected by the program (openssl is a prime suspect)

**Reply by uwehuberKEB** [reply]
I added the openssl library
![image](https://github.com/aprimadi/influxdb2/assets/76032392/ce898f7d-5bc0-44e6-8401-40b7d4bab6ff)
it still doesn't work
Unfortunately I added the line "Client builder created" after the screenshot of the Segmentation failed.
This screenshot has the line

**Reply by Boudewijn26** [reply]
Is it the `builder.build()` that's failing or are you unwrapping the error in the `match`?

I'd also recommend looking into why the trace isn't being logged. Maybe you've got a custom panic handler or maybe the ENV from your dockerfile isn't making its way into the program

**Reply by uwehuberKEB** [reply]
in the meantime I was able to get a stacktrace from the gdb debugger
New LWP 1902]
influxwriter started!
Redis connection established
Try to establish Influx connection
Client Builder created!

Thread 1 "influxwriter" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007f73c7bec7e7 in openssl::ssl::SslMethod::tls () at src/ssl/mod.rs:346
#2  0x00007f73c7be5f72 in native_tls::imp::TlsConnector::new (builder=0x7fff154e0800) at src/imp/openssl.rs:280
#3  0x00007f73c7be47c3 in native_tls::TlsConnectorBuilder::build (self=0x7fff154e0800) at src/lib.rs:446
**#4  0x00007f73c7a87d51 in reqwest::connect::Connector::new_default_tls<core::option::Option<core::net::ip_addr::IpAddr>> (http=..., t**ls=..., 
    proxies=Arc(strong=2, weak=0) = {...}, user_agent=<error reading variable: Cannot access memory at address 0xc9>, 
    local_addr=<error reading variable: Cannot access memory at address 0x3ba>, nodelay=true, tls_info=false) at src/connect.rs:93****
#5  0x00007f73c7a7f8d7 in reqwest::async_impl::client::ClientBuilder::build (self=...) at src/async_impl/client.rs:409
#6  0x00007f73c79ef275 in influxdb2::ClientBuilder::build (self=...) at src/lib.rs:279
#7  0x00007f73c79c523b in influxwriter::main::{async_block#0} () at src/main.rs:28
#8  0x00007f73c79cd734 in tokio::runtime::park::{impl#4}::block_on::{closure#0}<influxwriter::main::{async_block_env#0}> ()
    at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.36.0/src/runtime/park.rs:281
#9  0x00007f73c79cd575 in tokio::runtime::coop::with_budget<core::task::poll::Poll<()>, tokio::runtime::park::{impl#4}::block_on::{closure_env#0}<influxwriter::main::{async_block_env#0}>> (budget=..., f=...) at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.36.0/src/runtime/coop.rs:107
#10 tokio::runtime::coop::budget<core::task::poll::Poll<()>, tokio::runtime::park::{impl#4}::block_on::{closure_env#0}<influxwriter::main::{async_block_env#0}>> (f=...) at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.36.0/src/runtime/coop.rs:73
#11 tokio::runtime::park::CachedParkThread::block_on<influxwriter::main::{async_block_env#0}> (self=0x7fff154e3e77, f=...)
    at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.36.0/src/runtime/park.rs:281
#12 0x00007f73c79d66a5 in tokio::runtime::context::blocking::BlockingRegionGuard::block_on<influxwriter::main::{async_block_env#0}> (self=0x7fff154e5110, 
    f=...) at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.36.0/src/runtime/context/blocking.rs:66
#13 0x00007f73c79c6958 in tokio::runtime::scheduler::multi_thread::{impl#0}::block_on::{closure#0}<influxwriter::main::{async_block_env#0}> (
    blocking=0x7fff154e5110) at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.36.0/src/runtime/scheduler/multi_thread/mod.rs:87
#14 0x00007f73c79ce6dc in tokio::runtime::context::runtime::enter_runtime<tokio::runtime::scheduler::multi_thread::{impl#0}::block_on::{closure_env#0}<influxwr
--Type <RET> for more, q to quit, c to continue without paging--c
iter::main::{async_block_env#0}>, ()> (handle=0x7fff154e7f30, allow_block_in_place=true, f=...)
    at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.36.0/src/runtime/context/runtime.rs:65
#15 0x00007f73c79c6913 in tokio::runtime::scheduler::multi_thread::MultiThread::block_on<influxwriter::main::{async_block_env#0}> (self=0x7fff154e7f08, 
    handle=0x7fff154e7f30, future=...)
    at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.36.0/src/runtime/scheduler/multi_thread/mod.rs:86
#16 0x00007f73c79d691a in tokio::runtime::runtime::Runtime::block_on<influxwriter::main::{async_block_env#0}> (self=0x7fff154e7f00, future=...)
    at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.36.0/src/runtime/runtime.rs:350
#17 0x00007f73c79d64ec in influxwriter::main () at src/main.rs:29
(gdb) 

It seems that in line #4 is the memory problem.  reqwest::connect::Connector::new_default_tls 
Does this help?

**Reply by Boudewijn26** [reply]
It gives some direction. I'd first recommend checking if a `curl` from inside the container is able to reach the influx IP address. I've seen docker networks break by eg. reloading IP tables.

I also find it interesting that reqwest uses a tls connector. Are you connecting through https? Your code snippet uses http. That could point to reqwest missing some system library. You could try using rustls instead by changing the crate features or it might be a case of installing the right certificate inside the docker container

**Reply by uwehuberKEB** [reply, marked as answer]
Finally I found the problem. Thanks a lot for your help. The final hint came from the unsafe code in the debugger.
![image](https://github.com/aprimadi/influxdb2/assets/76032392/b468420d-c8de-41dd-9980-b002f2f5de9e)
I found the solution [here](https://users.rust-lang.org/t/sigsegv-with-program-linked-against-openssl-in-an-alpine-container/52172)
The problem seems to be that the x86_64-unknown-linux-musl target statically links to MUSL by default, but Alpine uses a dynamically linked MUSL.  Since you're dynamically linking to Alpine's OpenSSL, you're ending up with 2 MUSLs.
The entry 
ENV RUSTFLAGS="-C target-feature=-crt-static"
in the dockerfile was the solution.
